File: cpp/17. Letter Combinations of a Phone Number.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;
typedef vector<string> vs;

class Solution {
public:
    vs ans;
    string mapping[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    void dfs(int i, string &path, string &digits) {
        if(i==digits.size()) {
            ans.push_back(path);
            return;
        }
        int digit = digits[i] - '0';
        for(char ch : mapping[digit]) {
            path.push_back(ch);
            dfs(i+1, path, digits);
            path.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        if(digits.size()==0) return ans;
        string path;
        dfs(0, path, digits);
        return ans;
    }
};

void printAns(vs ans) {
    for(auto str : ans) {
        cout << str << " ";
    }
    cout << endl;
}

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; string digits; vs ans;

    solution = Solution();
    digits = "23";
    ans = solution.letterCombinations(digits);
    printAns(ans);

    solution = Solution();
    digits = "";
    ans = solution.letterCombinations(digits);
    printAns(ans);

    solution = Solution();
    digits = "2";
    ans = solution.letterCombinations(digits);
    printAns(ans);
}

File: python/17-Letter Combinations of a Phone Number.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        "Backtracking, Time: O(4^n), Space: O(n)"
        if digits == '': return []
        table = {
            '2': ['a', 'b', 'c'],
            '3': ['d', 'e', 'f'],
            '4': ['g', 'h', 'i'],
            '5': ['j', 'k', 'l'],
            '6': ['m', 'n', 'o'],
            '7': ['p', 'q', 'r', 's'],
            '8': ['t', 'u', 'v'],
            '9': ['w', 'x', 'y', 'z'],
        }
        results = []
        def dfs(i, path):
            if i == len(digits):
                results.append(path)
                return
            candidates = table[digits[i]]
            for candidate in candidates:
                dfs(i+1, path+candidate)
        dfs(0, '')
        return results

digits = "23"
result = Solution().letterCombinations(digits)
print(result) # ["ad","ae","af","bd","be","bf","cd","ce","cf"]

digits = ""
result = Solution().letterCombinations(digits)
print(result) # []

digits = "2"
result = Solution().letterCombinations(digits)
print(result) # ["a","b","c"]